sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,792 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/rpcutil"
+
+	pbempty "github.com/golang/protobuf/ptypes/empty"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v2/proto/go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ProviderMain is the typical entrypoint for a resource provider plugin. Using it isn't required but can cut down
+// significantly on the amount of boilerplate necessary to fire up a new resource provider.
+func ProviderMain(name, version string, args ProviderArgs) error {
```

Comment:
I'd originally wanted to put this in it's own `provider` package and name this function `Main` -- but then I wouldn't have been able to access some of the private details in `pulumi` that are needed for `Construct`'s implementation. Definitely open to alternative names for this.

Reply:
It is a little unfortunate that this all ends up in the top-level package - especially since some concepts overlap a bit with existing concepts with the same or similar names (like `Provider`).  Do we have an option to expose a smaller new surface area publicly here and then the rest of the API in a child package?  Not a blocker though if this is sufficiently unnatural.

Reply:
I'm going to noodle on this a bit this week to see what we can do.

Reply:
Two things: 1) The APIs are now exposed in the `pulumi/provider` package, rather than the top-level `pulumi` package and 2) the API has been whittled down to no longer expose a provider API. Instead, there is a `Construct` function that adapts the gRPC request/response to/from the Pulumi Go programming model, to be used in conjunction with the existing provider helper we have in https://github.com/pulumi/pulumi/tree/master/pkg/resource/provider.

I've also added a convenience function `ComponentMain` in the `pkg/resource/provider` package, that can be used to implement a minimal resource provider just for components, so you don't have to write all the other resource provider boilerplate just for `Construct`. I have a draft PR of the go-boilerplate repo that demonstrates its use: https://github.com/pulumi/pulumi-component-provider-go-boilerplate/pull/3

```go
import (
	"github.com/pulumi/pulumi/pkg/v3/resource/provider"
	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
)

// Serve launches the gRPC server for the resource provider.
func Serve(providerName, version string, schema []byte) {
	// Start gRPC service.
	if err := provider.ComponentMain(providerName, version, schema, construct); err != nil {
		cmdutil.ExitError(err.Error())
	}
}
```

```go
import (
	"github.com/pkg/errors"

	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi/provider"
)

func construct(ctx *pulumi.Context, typ, name string, inputs provider.ConstructInputs,
	options pulumi.ResourceOption) (*provider.ConstructResult, error) {
	// TODO: Add support for additional component resources here.
	switch typ {
	case "xyz:index:StaticPage":
		return constructStaticPage(ctx, name, inputs, options)
	default:
		return nil, errors.Errorf("unknown resource type %s", typ)
	}
}

// constructStaticPage is an implementation of Construct for the example StaticPage component.
// It demonstrates converting the raw ConstructInputs to the component's args struct, creating
// the component, and returning its URN and state (outputs).
func constructStaticPage(ctx *pulumi.Context, name string, inputs provider.ConstructInputs,
	options pulumi.ResourceOption) (*provider.ConstructResult, error) {

	// Copy the raw inputs to StaticPageArgs. `inputs.SetArgs` uses the types and `pulumi:` tags
	// on the struct's fields to convert the raw values to the appropriate Input types.
	args := &StaticPageArgs{}
	if err := inputs.SetArgs(args); err != nil {
		return nil, errors.Wrap(err, "setting args")
	}

	// Create the component resource.
	staticPage, err := NewStaticPage(ctx, name, args, options)
	if err != nil {
		return nil, errors.Wrap(err, "creating component")
	}

	// Return the component resource's URN and state. `NewConstructResult` automatically sets the
	// ConstructResult's state based on resource struct fields tagged with `pulumi:` tags with a value
	// that is convertible to `pulumi.Input`.
	return provider.NewConstructResult(staticPage)
}
```

I've also included a `NewConstructResult` helper along the lines of your suggestion, that creates the result with the URN and state based on the outputs defined on the component.